bin/convert_ismapper_to_vcf.py

```diff
@@ -0,0 +1,111 @@
+#!/usr/bin/env python3
+
+import argparse
+import csv
+from Bio import SeqIO
+
+
+
+# Define the VCF header
+vcf_header = """##fileformat=VCFv4.2
+##source=ISMapper
+##INFO=<ID=SVTYPE,Number=1,Type=String,Description="Type of structural variant">
+#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO
+"""
+
+# Function to read the reference genome
+def read_reference_genome(reference_file):
+    reference_sequences = {}
+    reference_sequence_length = 0
+    with open(reference_file, 'r') as ref_file:
+        # Read the reference genome using Biopython
+        for record in SeqIO.parse(ref_file, "fasta"):
+            reference_sequences[record.id] = record.seq
+            reference_sequence_length = len(record.seq)
+    return reference_sequences, reference_sequence_length
+
+# Function to read the transposable element information
+def read_transposable_elements(te_file):
+    te_info = {}
+    with open(te_file, 'r') as te_file:
+        # Read the multifasta using Biopython
+        for record in SeqIO.parse(te_file, "fasta"):
+            te_name = record.id
+            te_info[te_name] = len(record.seq)
+
+        # NOTE: Renable in case we need to dump this on disk.
+        # with open("temp_transposable_elements.csv", mode='w', newline='') as file:
+        #     writer = csv.writer(file)
+        #     writer.writerow(["name", "length"])
+        #     for te_name, te_length in te_info.items():
+        #         writer.writerow([te_name, te_length])
+    return te_info
+
+# Function to convert ISMapper data to VCF format
+def convert_is_mapper_to_vcf(is_mapper_file, vcf_file, reference_sequences, te_info):
+    with open(is_mapper_file, 'r') as infile_is_mapper, open(vcf_file, 'w') as outfile:
+        # Write the VCF header
+        outfile.write(vcf_header)
+
+        # Read the ISMapper TSV file
+        reader_is_mapper = csv.DictReader(infile_is_mapper, delimiter='\t')
+
+        for row in reader_is_mapper:
+            chrom = "NC-000962-3-H37Rv"  # Using the specified chromosome
+            pos = int(row['x'])  # Convert position to integer
+            region_id = row['region']
+            ref = reference_sequences[chrom][pos - 1]  # Extract the reference allele from the reference sequence
+            orientation = row['orientation']
+            te_name = row['transposable_element']
```

Comment:
This is currently the problematic code since in my experiments, the IS_Mapper txt file didn't produce this field.

PFA the outputs for one of the samples 

[ismapper.zip](https://github.com/user-attachments/files/16227565/ismapper.zip)

Reply:
sent new code!

Reply:
@vrennie , could you please try with the following files and script, using the data mentioned in https://github.com/TORCH-Consortium/MAGMA/pull/209#discussion_r1677134386 ? 

With this specific version, the generated VCF is not as per `tb-profiler` expectation

NOTE: Please just remove `txt` from the file suffixes.

[convert_ismapper_to_vcf.py.txt](https://github.com/user-attachments/files/16232768/convert_ismapper_to_vcf.py.txt)


[IS_Queries_Mtb.multi.fasta.txt](https://github.com/user-attachments/files/16232783/IS_Queries_Mtb.multi.fasta.txt)

You can use the script as shown here 

``` 
convert_ismapper_to_vcf.py \
    --is_mapper_file ismapper/IS6110/K059727_S61__NC_000962.3_table.txt \
    --reference_file NC-000962-3-H37Rv.fa \
    --query_file IS_Queries_Mtb.multi.fasta \
    --output_vcf_file output.vcf
```

Reply:
<img width="821" alt="image" src="https://github.com/user-attachments/assets/87a323ac-8ed6-475b-bb78-465d2fb3e7b5">

Reply:
@abhi18av did you get the same error?

Reply:
Ah, I made a typo in the script `qeury_file` - please just tweak it 🙏

Reply:
ok that works, but TBprofiler doesn't accept this output you say? @abhi18av 
[output.vcf.txt](https://github.com/user-attachments/files/16240622/output.vcf.txt)

Reply:
Exactly, when I used this file with our `tb-profiler` container it throws an error 

```
sh-5.2# tb-profiler profile --vcf output.vcf.txt 

[09:22:26] INFO     Using ref file: /usr/local/share/tbprofiler//tbdb.fasta                                                db.py:587
           INFO     Using gff file: /usr/local/share/tbprofiler//tbdb.gff                                                  db.py:587
           INFO     Using bed file: /usr/local/share/tbprofiler//tbdb.bed                                                  db.py:587
           INFO     Using json_db file: /usr/local/share/tbprofiler//tbdb.dr.json                                          db.py:587
           INFO     Using variables file: /usr/local/share/tbprofiler//tbdb.variables.json                                 db.py:587
           INFO     Using spoligotype_spacers file: /usr/local/share/tbprofiler//tbdb.spoligotype_spacers.txt              db.py:587
           INFO     Using spoligotype_annotations file: /usr/local/share/tbprofiler//tbdb.spoligotype_list.csv             db.py:587
           INFO     Using bedmask file: /usr/local/share/tbprofiler//tbdb.mask.bed                                         db.py:587
           INFO     Using barcode file: /usr/local/share/tbprofiler//tbdb.barcode.bed                                      db.py:587
           INFO     Using rules file: /usr/local/share/tbprofiler//tbdb.rules.txt                                          db.py:587
           ERROR    index: "output.vcf.txt" is in a format that cannot be usefully indexed                              utils.py:485
                                                                                                                                    
Traceback (most recent call last):
  File "/usr/local/bin/tb-profiler", line 587, in <module>
    args.func(args)
  File "/usr/local/bin/tb-profiler", line 95, in main_profile
    nsamples = len(tbp.get_vcf_samples(args.vcf))
  File "/usr/local/lib/python3.10/site-packages/tbprofiler/utils.py", line 18, in get_vcf_samples
    vcf = Vcf(vcf_file)
  File "/usr/local/lib/python3.10/site-packages/pathogenprofiler/vcf.py", line 53, in __init__
    tabix(filename,threads)
  File "/usr/local/lib/python3.10/site-packages/pathogenprofiler/utils.py", line 671, in tabix
    run_cmd(cmd)
  File "/usr/local/lib/python3.10/site-packages/pathogenprofiler/utils.py", line 486, in run_cmd
    raise ValueError("Command Failed:\n%s\nstderr:\n%s" % (cmd,result.stderr.decode()))
ValueError: Command Failed:
/bin/bash -c set -o pipefail; bcftools index --threads 1 -ft output.vcf.txt
stderr:
index: "output.vcf.txt" is in a format that cannot be usefully indexed

Cleaning up after failed run
           ERROR                                                                                                      tb-profiler:59
                                                                                                                                    
                    ################################# ERROR #######################################                                 
                                                                                                                                    
                    This run has failed. Please check all arguments and make sure all input files                                   
                    exist. If no solution is found, please open up an issue at                                                      
                    https://github.com/jodyphelan/TBProfiler/issues/new and paste or attach the                                     
                    contents of the error log (tbprofiler.errlog.txt)                                                               
                                                                                                                                    
                    ###############################################################################  
```

Reply:
Ah here again it needs the bcftools compression command I passed previously

Reply:
also the vcf is vcf.txt here so not ideal I think

Reply:
> Ah here again it needs the bcftools compression command I passed previously


Silly me 🤦 , yes once I used `bcftools`, it worked 🎉